xof.md

```diff
@@ -0,0 +1,206 @@
+
+At a minimum an XOF needs to support the following pseudo-code
+```
+xof = xof.new();
+xof.absorb(bytes1);
+xof.absorb(bytes2);
+xof.finalize();
+out1 = xof.squeeze(10);
+out2 = xof.squeeze(1000);
+xof.reset();
+xof.assorb(bytes);
+```
+### Rules
+
+- absorb can be called multiple times
+- finalize ends the absorb process (by adding padding bytes and doing a final absorb).
+  absorb must not be called once the finalize is done unless a reset happens.
+- finalize may be done as part of the first squeeze operation
+- squeeze can be called multiple times.
+
+OpenSSL XOF Requirements
+------------------------
+
+The current OpenSSL implementation of XOF only supports a single call to squeeze.
+The assumption exists in both the high level call to EVP_DigestFinalXOF() as well
+as in the lower level SHA3_squeeze() operation (Of which there is a generic c version,
+as well as assembler code for different platforms).
+
+A decision has to be made as to whether a new API is required, as well as considering
+how the change may affect existing applications.
+The changes introduced should have a minimal affect on other related functions that
+share the same code (e.g SHAKE and SHA3 share functionality).
+Older providers that have not been updated to support this change should produce an
+error if a newer core is used that supports multiple squeeze operations.
+
+API Discussion of Squeeze
+-------------------------
+
+### Squeeze
+
+Currently EVP_DigestFinalXOF() uses a flag to check that it is only invoked once.
+It returns an error if called more than once. When initially written it also did a 
+reset, but that code was removed as it was deemed to be incorrect.
+
+If we remove the flag check, then the core code will potentially call low level squeeze code
+in a older provider that does not handle returning correct data for multiple calls.
+To counter this the provider needs a mechanism to indicate that multiple calls are allowed.
+This could just be a new gettable flag (having a separate provider function should not be necessary).
+
+#### Proposal 1
+
+Change EVP_DigestFinalXOF(ctx, out, outlen) to handle multiple calls.
+Possibly have EVP_DigestSqueeze() just as an alias method?
+Changing the code at this level should be a simple matter of removing the flag check.
+
+##### Pros
+
+  - New API is not required
+
+##### Cons
+
+  - Final seems like a strange name to call multiple times.
+
+#### Proposal 2
+
+Keep EVP_DigestFinalXOF() as a one shot function and create a new API to handle the
+multi squeeze case e.g.
+```
+EVP_DigestSqueeze(ctx, out, outlen).
+```
+
+##### Pros
+
+  - Seems like a better name.
+
+##### Cons
+
+  - Adds an extra API.
+  - The interaction between the 2 API's needs to be clearly documented.
+     - A call to EVP_DigestSqueeze() after EVP_DigestFinalXOF() would fail.
+     - A call to EVP_DigestFinalXOF() after the EVP_DigestSqueeze() would fail? Is this confusing.
```

Comment:
It definitely needs to fail as finalization cannot be done multiple times.

Reply:
Perhaps creates a state transition diagram for XOF digests for comments?

`doc/life-cycles/digest.dot` would be a good basis to start from.  Generated images are in `doc/man7/img/` but this will require a new one I suspect rather than updating digests.dot